clippy_lints/src/disallowed_types.rs

```diff
@@ -51,43 +54,139 @@ declare_clippy_lint! {
     /// // A similar type that is allowed by the config
     /// use std::collections::HashMap;
     /// ```
+    ///
+    /// Disallowed profiles can scope lists to specific modules:
+    /// ```toml
+    /// [profiles.forward_pass]
+    /// disallowed-types = [{ path = "crate::buffers::HostBuffer", reason = "Prefer device buffers in forward computations" }]
+    /// ```
+    ///
+    /// ```rust,ignore
+    /// #[clippy::disallowed_profile("forward_pass")]
+    /// fn forward_step(buffer: crate::buffers::DeviceBuffer) { /* ... */ }
+    /// ```
     #[clippy::version = "1.55.0"]
     pub DISALLOWED_TYPES,
     style,
     "use of disallowed types"
 }
 
-pub struct DisallowedTypes {
+struct TypeLookup {
     def_ids: DefIdMap<(&'static str, &'static DisallowedPath)>,
     prim_tys: FxHashMap<PrimTy, (&'static str, &'static DisallowedPath)>,
 }
 
+impl TypeLookup {
+    fn from_config(tcx: TyCtxt<'_>, paths: &'static [DisallowedPath]) -> Self {
+        let (def_ids, prim_tys) = create_disallowed_map(tcx, paths, PathNS::Type, def_kind_predicate, "type", true);
+        Self { def_ids, prim_tys }
+    }
+
+    fn find(&self, res: &Res) -> Option<(&'static str, &'static DisallowedPath)> {
+        match res {
+            Res::Def(_, did) => self.def_ids.get(did).copied(),
+            Res::PrimTy(prim) => self.prim_tys.get(prim).copied(),
+            _ => None,
+        }
+    }
+}
+
+pub struct DisallowedTypes {
+    default: TypeLookup,
+    profiles: FxHashMap<Symbol, TypeLookup>,
+    known_profiles: FxHashSet<Symbol>,
```

Comment:
Why is `profiles` different from `known_profiles`? Is there a difference between these two? Seems that `known_profiles` is just a set from the keys of `profiles`.